Implement multi-selecting objects

wjwilimowski opened the pull request:
Added the ability to multi-select `ItemInstance` objects by Ctrl-clicking
- Move all selected objects
- Rotate all selected objects
- Rotate all selected objects as a group (preserving relative position) by holding down Alt
- Delete selected objects
- Copy & paste selected objects
- Undo for all above operations

Other changes
- Added TombLib.Tests project
- Added tests for parts of multi-select functionality

wjwilimowski commented:
Hey! I thought a feature like this would add a lot of convenience :)

I've implemented all the operations I could think of, but please tell me if I've missed something that would leave an awkward gap in the functionality - I'm worried about missing something.

wjwilimowski commented:
There is one case where I know the behavior differs from just working with a single object selected - when you use "Rotate room" with an object selected, the object is still selected after the rotation, but in case you had multiple objects selected, they are not selected anymore after you rotate rooms. This is because `Room.AnyObjects` does not return the `ObjectGroup`. I figured it's not worth implementing it, but I can fix it if we want to go for consistency.

MontyTRC89 commented:
Hi! thank you :-)
right now editor has received more releases and i will update that repo. I will inspect your code and see if it can work with newest releases because it's a feature we miss a lot, multiselect :-D

wjwilimowski commented:
> Hi! thank you :-)
> right now editor has received more releases and i will update that repo. I will inspect your code and see if it can work with newest releases because it's a feature we miss a lot, multiselect :-D

I'll rebase onto the new changes when I'm able to find the time :)

MontyTRC89 commented:
Would you want to join the official team? We need new features specific for TombEngine, I am developing them but leading two projects it's time intensive. We need fresh minds :-D

Lwmte commented:
Hi, I just wanted to say that unfortunately this pull request breaks VS2015 compatibility totally. There are several people at TE dev team who still use VS2015 or VS2017, so there may be problems with using this code on their systems.

Lwmte commented:
Also another thing I noticed - it seems that currently you can only multiselect moveables/statics but not "service objects" such as cameras, sinks, flybys, sound sources etc?

wjwilimowski commented:
> Hi, I just wanted to say that unfortunately this pull request breaks VS2015 compatibility totally. There are several people at TE dev team who still use VS2015 or VS2017, so there may be problems with using this code on their systems.

Oh, sorry, I didn't think about that - I'm used to those new language features at work. I'll fix the compatibility. Maybe I can try adding some editor config file as well, which will make IDEs flag incompatible language features as build errors - I'll see if it can be done.

> Also another thing I noticed - it seems that currently you can only multiselect moveables/statics but not "service objects" such as cameras, sinks, flybys, sound sources etc?

Yep, that's correct - it's just as deep as I've gone on my dive into the class hierarchy. I'll investigate how they can be added

wjwilimowski commented:
> Would you want to join the official team? We need new features specific for TombEngine, I am developing them but leading two projects it's time intensive. We need fresh minds :-D

I'm very happy to receive this invitation :) However I'm afraid I wouldn't be able to reliably schedule some time for this work, and I would just end up feeling bad about not living up to commitments and coming off as unreliable - I'll have to decline.

Lwmte commented:
> Oh, sorry, I didn't think about that - I'm used to those new language features at work. I'll fix the compatibility. Maybe I can try adding some editor config file as well, which will make IDEs flag incompatible language features as build errors - I'll see if it can be done.

I am not sure what do you mean, since VS2015 already marks incompatible code as errors.

Regarding VS2017+ code features, it wouldnt be critical if you worked on a side project (e.g. TIDE), because TIDE already uses a lot of VS2019 features, but it is developed by a single developer, so others can just disable this project and still will be able to compile other stuff on VS2015. Since TE is a main program in the toolchain, sadly it should be compatible with VS2015 since there are at least 2 devs who need that (me and ChocolateFan).

wjwilimowski commented:
@Lwmte - It seems I have fixed all the compatibility issues I've introduced - I checked and I can build the solution and run it in VS2015.

I've also added `<LangVersion>6</LangVersion>` to the relevant `csproj` files - now any IDE will treat using language features higher than C# 6.0 as build errors. This should prevent people from making the same mistake :)

Here's what it looks like in VS2019

![image](https://user-images.githubusercontent.com/11385821/127062132-b2c94fdc-98df-4ef5-8c16-c74233086e7e.png)

And in Rider

![image](https://user-images.githubusercontent.com/11385821/127062457-dfe7e06c-3650-4c56-a007-6500ae4e67c6.png)

Lwmte commented:
Wow, thank you for that! Indeed, it frequently happens when someone tries to use new C# features and then have to rollback.

wjwilimowski commented:
Hey guys, I've made a few changes.

- Multiselect now supports all `PositionBasedObjectInstance`
- Object group now has a tooltip like all other selected objects
- Improved selection process
  - Ctrl-clicking an object which is already part of the group deselects just that object
  - Ctrl-clicking on objects that can't be multi-selected does nothing (prevents accidentally deselecting)
  - You can now start multi-selecting by ctrl-clicking the first object, which was not possible before
  
I've also tried to add support for Imported Geometry and Static Meshes, but I don't even know how to use that feature, so I wasn't able to test it 🤣 Could you please explain how to use those? I've based the implementation on what I've done for `MoveableInstance` since it seems to be a similar case.

wjwilimowski commented on TombEditor/Controls/PanelRendering3D.cs:
```diff
@@ -2427,6 +2468,11 @@ private void DrawVolumes(Matrix4x4 viewProjection, Effect effect, List<VolumeIns
 
         private void DrawObjects(Matrix4x4 viewProjection, Effect effect, Room[] roomsWhoseObjectsToDraw, List<Text> textToDraw)
```
I have a suggestion here - could we rename this method to `DrawPlaceholders`? When I first investigated this code, it confused me that Moveable meshes are drawn outside this method. This way it will be clear that the method draws just the little boxes and spheres in case something does not have a mesh representation.

Lwmte replied:
Yeah, you are right, `DrawObjects` is mostly for rendering "service objects" as I call them. I will rename function when your PR is merged to avoid merge conflicts.

Lwmte commented:
I have tested after latest commits, and everything works okay on the first sight! Only issue I have noticed is when you deselect objects from group and only one object is left in group, it is not converted back to single object. Also in case "origin" object was deselected from group, gizmo stays on old position. I am not sure if it is feature or bug, because it may be useful.

Lwmte commented:
By the way, I saw that you have edited `PanelRendering3D.cs` in latest commits, namely text labels (`textToDraw.Add()` and stuff). There could be probably conflicts when Monty updates this repo, because I have simplified exactly this part of code to deduplicate it. Just a small note to know what's coming.
And congratulations on a great job again! :)

MontyTRC89 commented:
Hi! We are planning how to merge your PR in our private master. It's a delicate task so we'll think about it very carefully. But your change is very very good and it will be part of 1.4.

wjwilimowski commented:
@Lwmte 

> I have tested after latest commits, and everything works okay on the first sight! Only issue I have noticed is when you deselect objects from group and only one object is left in group, it is not converted back to single object. 

That's a good suggestion with turning the selection back into single object - I'll add it.

> Also in case "origin" object was deselected from group, gizmo stays on old position. I am not sure if it is feature or bug, because it may be useful.

Yeah, that was intentional on my part, because there's no good way to pick which object should become the next centerpoint. That makes me think it would be useful to to let the user move the gizmo for a multi-selection anywhere they want.
 
> You mean support for what? Both imported geometry and statics are also position-based object instances. Only difference for imported geometries is they can be rotated in all directions, not just horizontally (this is also valid for flyby cameras and volumes). But since ability to group rotate on other axis than X never could happen in real-life scenarios, I suggest you not to bother about that.

Sorry, didn't make myself clear - I meant the rendering code in `DrawImportedGeometry` and `DrawStatics`.

Lwmte commented:
To be honest, I don't see big difference in rendering code, it looks good to me. Only thing I am a bit concerned about is this part:

`var activeObjectGroup = _editor.SelectedObject as ObjectGroup;
            var isSelected = activeObjectGroup != null
                ? (Func<LightInstance, bool>)(light => activeObjectGroup.Contains(light))
                : light => light == _editor.SelectedObject;`

I don't fully understand this code and it looks unnecessary complex to me. In some cases `isSelected` is omitted and you directly check `activeObjectGroup != null && activeObjectGroup.Contains()`. It looks more clear to me. Anyway, both variations are copypasted in several places, I think you can move those to single helper function, like this one, and use it everywhere:

`bool IsSelected(PositionBasedObjectInstance item) 
{
      var activeObjectGroup = _editor.SelectedObject as ObjectGroup;
      return editor.SelectedObject == item || (activeObjectGroup != null && activeObjectGroup.Contains(item));
}`

Lwmte commented on TombLib/LevelData/Instances/ObjectInstance.cs:
```diff
@@ -225,6 +225,11 @@ public abstract class PositionBasedObjectInstance : ObjectInstance, ISpatial
     {
         public Vector3 Position { get; set; }
 
+        public virtual void SetPosition(Vector3 position)
```
Should we always use `SetPosition` now or there are specific cases only for ObjectGroup? I see there a possible confusion between old `Position` field and new `SetPostion()` method.

wjwilimowski replied:
Actually, after reading your comment I realized I can just make `SetPosition` protected, make `Position` a property with backing field `_position`, and call `SetPosition` from the setter. This way it's still overridable, but looks the same as it always has from the outside.

wjwilimowski replied:
Good suggestion, thanks :)

wjwilimowski commented:
> To be honest, I don't see big difference in rendering code, it looks good to me. Only thing I am a bit concerned about is this part:

The reason I'm doing it like this is because I wanted to get rid of checking `activeObjectGroup != null` all the time. If I create a function `isSelected(object o)` first, which is either `o == editor.SelectedObject` or `activeObjectGroup.Contains(o)` up front, I only check `activeObjectGroup != null` once.

Nevertheless, you're right that it looks ugly and it can be factored out for more clarity. I'll also look if I'm mixing both methods to make it consistent. I'll fix it :)

wjwilimowski commented:
Just leaving a comment to remember that copy-pasting an object group that contains `ImportedGeometry` is broken - will fix once I figure out what's going on

wjwilimowski commented:
> I have tested after latest commits, and everything works okay on the first sight! Only issue I have noticed is when you deselect objects from group and only one object is left in group, it is not converted back to single object.

I've fixed this behavior. That being said, it would have been useful if you wanted to rotate object A around a stationary object B. Select B, ctrl-click A, deselect B, spin A around B. I suppose this is a niche case though.

Lwmte commented:
> I found out that copying should be blocked for any `ImportedGeometry` where `Model == null`. Fixed for `ObjectGroup` copying.

Ahh, yes, sadly it is a hack to avoid exceptions. Imported geometry class was written by a guy who is gone from dev team and nobody can fix this code due to its complexity. Sorry for that :(

I will test your code again in a short time, I believe if no more issues will afloat, it's good to go.

wjwilimowski commented:
> Imported geometry class was written by a guy who is gone from dev team and nobody can fix this code due to its complexity.

Sounds like a challenge ;)

Lwmte commented:
I don't think that you should spend time on that, only if you have a lot of free time to invest in it... It's not the only occurence of such hacks in TE code, there are many of them sadly.

I did some additional tests and I can't find any major problems right now. Also I tried to merge your PR into current dev version, of course auto merge failed and git wrongly guessed many changes - biggest problem is `PanelRendering3D` class, I guess Monty will have to manually merge it. Let's wait on his comments (I think he should also test before merging).

Lwmte commented:
I have merged your code to dev repository, but found some critical bugs afterwards (also exist on this repository):

1) Try to cut or delete objects in group. You will see hanging gizmo without objects. Now if you try to move or rotate it, all the hell breaks loose with exceptions.
2) Sometimes pasting objects which were cut before in differnt places (mainly near room borders) also results in exceptions related to `PlaceObjectWithoutUpdate` function.

wjwilimowski commented:
@Lwmte - thanks for the detailed feedback! I'll take a look at it

Lwmte commented:
After further investigations, I realized that there is fundamental flaw in your approach.

Basically `ObjectGroup` instance is never added into a room when you do mouse selection. It stays only as selected object, it is "non-material" instance. However, when you copy-paste it, it is actually added to room as an object. That's why I have either exceptions when I am trying to fix undo-redo with copypaste or when I am trying to fix undo-redo with deletion.

I think that `ObjectGroup` should NEVER actually spill into a room, it should stay as "ephemeric" selected object, and that's all. I also tried to implement it, but I faced issues with item positioning (you use `SetPosition` method to somehow override child positions, and I can't properly set it).

Lwmte commented:
Another thing to note. When lights are moved/rotated together, room does not relight.

wjwilimowski commented:
Yeah, that is a problem, one of the base classes has Room on it but it doesn't really fit ObjectGroup. I'll have to look for a way to decouple it from Room

Lwmte commented:
Well, actually it's not a big problem if you still inherit from `PositionBasedObjectInstance` class, you can just ignore `Room` field for object group.

Lwmte commented:
Another issue I have found: if you "accumulate" group by cutting objects from different rooms and then pasting them in other rooms, accumulate there and copy again, undo stack tends to become corrupted and exception avalanche begins:

https://user-images.githubusercontent.com/3331699/129479045-0af8ca4b-0a35-4832-b396-60c4b273bc2f.mp4

This demo is from dev branch but same thing happens in this one.

Lwmte commented:
I have seemingly resolved all the issues with exception avalanche on the dev branch. All of them were related to `ObjectGroup` instance spilling into undo stack. I have removed your extensions to `TransformObjectUndoInstance` for object group and instead made an override for `PushObjectTransformed` function which converted object group into a list of children while pushing on stack. Therefore, all the bugs I initially reported are now somewhat resolved, but if you're working on the issue yourself, chances high you may resolve it in a less hacky way :)

Also, a bit of tease on what is now possible on dev branch thanks to your code:

https://user-images.githubusercontent.com/3331699/129497546-c6a1fc61-047e-4d56-a86a-1d88eedf09fa.mp4

wjwilimowski commented:
@Lwmte this is awesome!

Sorry for my limited input at the moment, I have a lot of things going on. Glad you were able to resolve the issues, and doubly glad that it was by removing special cases 😆

Lwmte commented:
No problem! I guess we will release new version with current code - our test team reports no major issues with it. After new release, code should be made available as public repo by Monty. If you will find time to review and finish it, it will be great. Thank you so much for this! :)
